include/nbl/video/utilities/CComputeBlit.h

```diff
@@ -0,0 +1,900 @@
+		case EF_R16_UNORM:
+		case EF_R16_SNORM:
+			return EF_R32_SFLOAT;
+
+		case EF_A2B10G10R10_UNORM_PACK32:
+		case EF_R8G8B8A8_UNORM:
+			return EF_R16G16B16A16_UNORM;
+
+		case EF_R8G8_UNORM:
+			return EF_R16G16_UNORM;
+
+		case EF_R8_UNORM:
+			return EF_R16_UNORM;
+
+		case EF_R8G8B8A8_SNORM:
+			return EF_R16G16B16A16_SNORM;
+
+		case EF_R8G8_SNORM:
+			return EF_R16G16_SNORM;
+
+		default:
+			return EF_UNKNOWN;
+		}
+	}
+
+private:
+	enum E_BLIT_TYPE : uint8_t
+	{
+		EBT_REGULAR = 0,
+		EBT_COVERAGE_ADJUSTMENT,
+		EBT_COUNT
+	};
+
+	core::smart_refctd_ptr<video::IGPUDescriptorSetLayout> m_blitDSLayout[EBT_COUNT];
+	core::smart_refctd_ptr<video::IGPUDescriptorSetLayout> m_kernelWeightsDSLayout;
+
+	core::smart_refctd_ptr<video::IGPUPipelineLayout> m_blitPipelineLayout[EBT_COUNT];
+	core::smart_refctd_ptr<video::IGPUPipelineLayout> m_coverageAdjustmentPipelineLayout;
+
+	core::smart_refctd_ptr<video::IGPUComputePipeline> m_alphaTestPipelines[MaxAlphaBinCount / MinAlphaBinCount][asset::IImage::ET_COUNT] = { nullptr };
+
+	struct SNormalizationCacheKey
+	{
+		asset::IImage::E_TYPE imageType;
+		uint32_t alphaBinCount;
+		asset::E_FORMAT outFormat;
+
+		inline bool operator==(const SNormalizationCacheKey& other) const
+		{
+			return (imageType == other.imageType) && (alphaBinCount == other.alphaBinCount) && (outFormat == other.outFormat);
+		}
+	};
+	struct SNormalizationCacheHash
+	{
+		inline size_t operator() (const SNormalizationCacheKey& key) const
+		{
+			return
+				std::hash<decltype(key.imageType)>{}(key.imageType) ^
+				std::hash<decltype(key.alphaBinCount)>{}(key.alphaBinCount) ^
+				std::hash<decltype(key.outFormat)>{}(key.outFormat);
+		}
+	};
+	core::unordered_map<SNormalizationCacheKey, core::smart_refctd_ptr<video::IGPUComputePipeline>, SNormalizationCacheHash> m_normalizationPipelines;
+
+	struct SBlitCacheKey
+	{
+		uint32_t wgSize;
+		asset::IImage::E_TYPE imageType;
+		uint32_t alphaBinCount;
+		asset::E_FORMAT outFormat;
+		uint32_t smemSize;
+		bool coverageAdjustment;
+
+		inline bool operator==(const SBlitCacheKey& other) const
+		{
+			return (wgSize == other.wgSize) && (imageType == other.imageType) && (alphaBinCount == other.alphaBinCount) && (outFormat == other.outFormat)
+				&& (smemSize == other.smemSize) && (coverageAdjustment == other.coverageAdjustment);
+		}
+	};
+	struct SBlitCacheHash
+	{
+		inline size_t operator()(const SBlitCacheKey& key) const
+		{
+			return
+				std::hash<decltype(key.wgSize)>{}(key.wgSize) ^
+				std::hash<decltype(key.imageType)>{}(key.imageType) ^
+				std::hash<decltype(key.alphaBinCount)>{}(key.alphaBinCount) ^
+				std::hash<decltype(key.outFormat)>{}(key.outFormat) ^
+				std::hash<decltype(key.smemSize)>{}(key.smemSize) ^
+				std::hash<decltype(key.coverageAdjustment)>{}(key.coverageAdjustment);
+		}
+	};
+	core::unordered_map<SBlitCacheKey, core::smart_refctd_ptr<video::IGPUComputePipeline>, SBlitCacheHash> m_blitPipelines;
+
+	uint32_t m_availableSharedMemory;
```

Comment:
either you make this variable a `const` or don't keep it around, can't have both.

Reply:
But then the user won't be able to call [`setAvailableSharedMemory`](https://github.com/Devsh-Graphics-Programming/Nabla/blob/5374db1c4b601e8a11d057ea359cb45440f0fa42/include/nbl/video/utilities/CComputeBlit.h#L78) at a later time. I wrote it with this in mind.

Reply:
If we don't want this, then I should just inline the code [here](https://github.com/Devsh-Graphics-Programming/Nabla/blob/5374db1c4b601e8a11d057ea359cb45440f0fa42/include/nbl/video/utilities/CComputeBlit.h#L34)?

Reply:
either smem needs to come from constructor and never change, or accept the smem size when creating individual pipelines and make your cache work with that

Reply:
Okay then I choose the former
> smem needs to come from constructor and never change